bundles/org.openhab.binding.yamahamusiccast/src/main/java/org/openhab/binding/yamahamusiccast/internal/YamahaMusiccastBridgeHandler.java

```diff
@@ -0,0 +1,143 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.yamahamusiccast.internal.dto.UdpMessage;
+import org.openhab.core.common.NamedThreadFactory;
+import org.openhab.core.thing.*;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link YamahaMusiccastBridgeHandler} is responsible for dispatching UDP events to linked Things.
+ *
+ * @author Lennert Coopman - Initial contribution
+ */
+@NonNullByDefault
+public class YamahaMusiccastBridgeHandler extends BaseBridgeHandler {
+    private Gson gson = new Gson();
+    private final Logger logger = LoggerFactory.getLogger(YamahaMusiccastBridgeHandler.class);
+    private @Nullable ExecutorService executor;
+    private @Nullable Future<?> eventListenerJob;
+    private static final int UDP_PORT = 41100;
+    private static final int SOCKET_TIMEOUT_MILLISECONDS = 3000;
+    private static final int BUFFER_SIZE = 5120;
+    private @Nullable DatagramSocket socket;
+
+    private void receivePackets() {
+        try {
+            DatagramSocket s = new DatagramSocket(null);
+            s.setSoTimeout(SOCKET_TIMEOUT_MILLISECONDS);
+            s.setReuseAddress(true);
+            InetSocketAddress address = new InetSocketAddress(UDP_PORT);
+            s.bind(address);
+            socket = s;
+            logger.debug("UDP Listener got socket on port {} with timeout {}", UDP_PORT, SOCKET_TIMEOUT_MILLISECONDS);
+        } catch (SocketException e) {
+            logger.debug("UDP Listener got SocketException: {}", e.getMessage(), e);
+            socket = null;
+            return;
+        }
+
+        DatagramPacket packet = new DatagramPacket(new byte[BUFFER_SIZE], BUFFER_SIZE);
+        while (socket != null) {
+            try {
+                socket.receive(packet);
+                String received = new String(packet.getData(), 0, packet.getLength());
+                String trackingID = UUID.randomUUID().toString().replace("-", "").substring(0, 32);
+                logger.debug("Received packet: {} (Tracking: {})", received, trackingID);
+                handleUDPEvent(received, trackingID);
+            } catch (SocketTimeoutException e) {
+                // Nothing to do on socket timeout
+            } catch (IOException e) {
+                logger.debug("UDP Listener got IOException waiting for datagram: {}", e.getMessage());
+                socket = null;
+            }
+        }
+        logger.debug("UDP Listener exiting");
+    }
+
+    public YamahaMusiccastBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.ONLINE);
+        if (eventListenerJob == null || eventListenerJob.isCancelled()) {
```

Comment:
To get rid of the compiler warning, you can store `eventListenerJob` to a local variable and do the null check on that.

Reply:
I just recompiled the binding with `mvn clean install` and the file `summary_report.html` doesn't show any remarks. What warning are you refering to?

Reply:
Compiler warnings don't show up in the summary_report.html. They are displayed in the output of `mvn clean install`.

Reply:
Beside that, I see two checkstyle warnings:
![grafik](https://user-images.githubusercontent.com/1894841/109712255-1909f980-7ba0-11eb-98f0-459fce1e9d9b.png)

Reply:
The checkstyle have been fixed.
However, I'm struggling with storing eventListenerJob to a local variable. Futures are a whole new concept to me.
I tried finding other examples but did not find anything.
Any suggestion/example to get me going with this? Thanks.

Reply:
I might have found a solution.
```
       Future<?> localEventlistenerJob = eventListenerJob;
        if (localEventlistenerJob == null || localEventlistenerJob.isCancelled()) {
            executor = Executors.newSingleThreadExecutor(new NamedThreadFactory("binding-yamahamusiccast"));
            localEventlistenerJob = executor.submit(this::receivePackets);
        }
```
It took me a while to figure it out.

Reply:
Looks good! Great that you figured it out yourself! The executor should be created only once. E.g. as a class' field. And the name of the executor should be the Thing UID.

Keep in mind that any executor must be shut-down by invoking `executor.shutdown()` when the Thing is disposed.

Reply:
OK, thx. Another challenge to place the executor somewhere else.
The `executor.shutdown()` is part of the dispose method.

Reply:
Found it (I think). Just pushed all changes with also a few new features.